### ISSUE TITLE
Problems with fcgi backend which doesn't support http2

### ISSUE THREAD

**Issue by akrus** [haproxy/haproxy#1232]
## Detailed description of the problem
We have fcgi-app as a backend. The app is based on Mono (hyperfastcgi4) and is pretty old, so it expects only HTTP/1.1 requests. When haproxy frontend accepts h2, it sends to the backend full URL in GET request:
GET https://svc.name/index.asp

And fails as hyperfastcgi passes the whole line there and cannot find the requested path as a result.
Using alpn http/1.1 fixes the problem, but it would be nice to keep h2 where possible (frontend is serving all the clients, not just this specific backend).

Currently there is no way to change this behavior and to force sending GET /index.asp with Host header.

## Expected behavior
An option should be available to work with http/1 backends.

## Steps to reproduce the behavior

1. Connect to URL with HTTP/2

## What is your configuration?

```
frontend services
  bind 0.0.0.0:80
  bind 0.0.0.0:443 ssl crt /usr/local/etc/haproxy/keys/ssl.pem alpn h2,http/1.1

  use_backend service_backend

backend service_backend
  use-fcgi-app mono
  server svc_fcgi svc_fcgi:9000 proto fcgi

fcgi-app mono
   log-stderr global
   option keep-conn
   option mpxs-conns
   set-param HTTP_CONTENT_TYPE %[req.hdr(Content-Type)]
   set-param HTTP_CONTENT_LENGTH %[req.hdr(Content-Length)]
   docroot /app
```

## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 2.3.9-1ppa1~bionic 2021/03/30 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2022.
Known bugs: http://www.haproxy.org/bugs/bugs-2.3.9.html
Running on: Linux 4.15.0-45-generic #48-Ubuntu SMP Tue Jan 29 16:28:13 UTC 2019 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-NwOlqk/haproxy-2.3.9=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=2).
Built with OpenSSL version : OpenSSL 1.1.1  11 Sep 2018
Running on OpenSSL version : OpenSSL 1.1.1  11 Sep 2018
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with the Prometheus exporter as a service
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.31 2018-02-12
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 7.5.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : prometheus-exporter
Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace
```

```
Linux svc1 4.15.0-45-generic #48-Ubuntu SMP Tue Jan 29 16:28:13 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by capflam** [comment]
I confirm the issue, I'm on it.

**Comment by capflam** [comment]
A fix was pushed in dev. However, note that it is not a real bug because absolute URI should be supported by H1 applications. And it is still possible for an H1 client to send an absolute URI. It is unusual but possible. In this case, HAProxy will not extract the path. It only works in H2 to H1 scenario, to conform to the habits of both worlds.

The patch will be backported soon. In the mean time, and to avoid any problem if an H1 client sends an absolute URI, you may replace the uri with a http-request rule. For instance : 
```
 http-request set-uri "%[pathq]"
```

**Comment by capflam** [comment]
Backported now. The fix will be shipped in the next 2.2 release (planned this week).
